sphinx gallery for docs

choldgraf opened the pull request:
This PR converts some of the documentation to examples using the sphinx-gallery plugin. This does a couple of things:

0. Converts many of the examples in `.rst` files (that were using `.. ipython::` blocks) into proper python files with `rst` embedded as comments
1. Generates a gallery of images showing off example outputs
2. Converts python files with embedded rST into a rendered HTML output
3. Creates jupyter notebooks for each example as well, to go along with the python file
4. Embeds examples in documentation stub pages
5. Makes (some) methods/classes/etc clickable links in the rendered examples.

This PR is still somewhat WIP, but this is a general idea of what it'd look like:

http://predictablynoisy.com/geopandas/gallery/index.html

And here's an example of the examples listed in an API stub page:

http://predictablynoisy.com/geopandas/api/_as_gen/geopandas.datasets.get_path.html#geopandas.datasets.get_path

Let me know what folks think and/or if this is worth me spending more time to refine. I think it'd be a nice step towards making the documentation more user-friendly! Comments welcome!

 Note: In general this does **not** change the content of any examples, it only converts them to python. I think that changing the example content itself should be done in another PR if folks want this.

jdmcbr commented:
Cool, I agree this pushes the documentation in a more user friendly direction.

jorisvandenbossche commented:
@choldgraf Sorry for the late reply. In general, I think it would be very cool to have an example gallery. My only reservation about this specific PR is that it seems you converted most pages of the user guide to example entries, while IMO they were not really meant/written as examples. And anyhow it is nice to have both user guide and more concrete examples in a gallery.

choldgraf commented:
Good point - the way that I've seen other packages do this is to have two sphinx gallery folders. One for `examples`, and another for `tutorials` or `guide` that is more descriptive, long-form content. The important thing is that whatever folder structure is used, they will all be runnable python scripts instead of `.rst` files. What would you think about something like that?

jorisvandenbossche commented:
But eg scikit-learn has a user guide with rst files, and then an example gallery (and links from api docs and user guide to example are provided). Same for matplotlib I think?

What is the advantage of writing your user guide as examples? (So people can download it as a notebook?)

It seems writing a user guide using python files is a bit more complex

jorisvandenbossche commented:
BTW, we also have a top-level 'examples' directory, we should look at converting those to gallery examples.

jorisvandenbossche commented on doc/download_example_data.py:
```diff
@@ -0,0 +1,9 @@
+path_data = os.path.join('..', 'gallery', 'data')
+path_zip = os.path.join(path_data, 'nybb_16a.zip')
+download(url, path_zip)
+download(url, path_data, zipfile=True)
```
This is not needed anymore. In the meantime, the nybb dataset is available from `geopandas.datasets.get_path('nybb')`

choldgraf commented:
actually on master matplotlib converted most of its user guide to a sphinx-gallery tutorials page :-)

http://matplotlib.org/devdocs/index.html

(full disclosure, I am the person that did that, so take my opinion with a grain of salt)

I think the biggest benefits with sphinx-gallery (or any gallery for that matter) are that:

* people in general respond to images much more strongly than text, so laying out viz in a visual matter is a natural way to present a high-level view of the material. 
* sphinx gallery has nice things like auto-embedding links to API documentation, auto-conversion into jupyter notebooks
* documentation for data analytics projects really shouldn't be text-heavy anyway, it should be code + viz heavy since that'll draw users into the content much more.
* more generally, everything is a runnable python script, so it makes it harder for docs to become outdated (since they'll have to be actually run in order for the documentation to get created...this was a huge problem with MPL).

That said, I totally understand where you're coming from with respect to having an rST-based user guide. sphinx-gallery is new enough that I don't know the community has found the obvious balance yet. I'll be flexible on this PR depending on what the community wants.

Also - I didn't realize there was an examples page...where is that link on the website?

jorisvandenbossche commented:
> Also - I didn't realize there was an examples page...where is that link on the website?

No website page, just in the repo: https://github.com/geopandas/geopandas/tree/master/examples But they are probably mainly outdated / maybe not relevant anymore. But therefore having those examples in a gallery will indeed make them better maintained (as they need to run). 
Eg https://github.com/geopandas/geopandas/blob/master/examples/choropleths.ipynb is maybe a good candidate to turn into a gallery example.

choldgraf commented:
ahhh yes - I do remember seeing that. I started trying to convert those a while back but it seemed like they were, in fact, out of date and I don't know the ins and outs of geopandas enough to quickly resolve the issues.

BTW, for another example of the `examples / tutorials` split, here's another project that I work with:

https://martinos.org/mne/dev/documentation.html

and here's the codebase:

https://github.com/mne-tools/mne-python

we've got a `tutorials` folder that contains all the tutorial python scripts, and then we insert those scripts into the documentation `.rst` files as links. That way we aren't really using a "gallery" for the tutorials, but we get all the sphinx-gallery function linking etc.

jorisvandenbossche commented:
Just to be sure (but I suppose the answer is no): sphinx gallery has no ability to deal with example files written in rst? (and converting those to python files to download instead of the other way around) Or as notebooks? 
In principle this could be possible, and gives the user more freedom to write in the format he/she likes, but still have the niceties of the gallery.

(it's just that I don't really like writing longer user guide pages in python files)

choldgraf commented:
makes sense - but I'm pretty sure the answer would be "no" from the SG team :-)

Though I do agree with you that it should be relatively straightforward to convert `.rst` to sphinx-gallery python scripts.

Part of the reason that sphinx-gallery was written is that the `pyplot` directive written for rST is notoriously buggy and difficult to optimize, improve, and maintain. I think their thought process is that trying to shove arbitrary executable python code into rST was clunkier than trying to shove rST into python code haha

jorisvandenbossche commented:
>  the pyplot directive written for rST is notoriously buggy and difficult to optimize, improve, and maintain. I think their thought process is that trying to shove arbitrary executable python code into rST was clunkier than trying to shove rST into python code

On the other hand, the ipython directive has served us very well in the pandas docs (all code examples are ran and kept up to date on each doc build). IMO this would be a nice format to write gallery example entries. But that is a discussion for somewhere else! :-)

(BTW, something else I 'dislike' about the gallery examples vs ipython directive is that the output (non-figure output) is more 'detached' from the code that produces it)

Now, in order to go forward with this for geopandas, may I propose the following (more conservative) approach?

- Let's start with adding the gallery 'infrastructure' but leave the existing user guide as is for now (including at least one example, eg the chloropleth one I linked to above). That can be in this PR or a new one.
- That way we can at least start adding new examples in the correct format (eg https://github.com/geopandas/geopandas/issues/392  would also be a perfect fit, as well as your cartopy example) in other PRs
- And we can still convert the user guide later to use the gallery as well (as we use it for the examples, we get more acquainted with it, and maybe better see the benefits you listed :-))

Would that be OK for you?

choldgraf commented:
> On the other hand, the ipython directive has served us very well in the pandas docs

For sure - it's just quite buggy and complex to develop :-P

>  something else I 'dislike' about the gallery examples vs ipython directive is that the output (non-figure output) is more 'detached' from the code that produces it

How so? It should be collected within blocks of the code...

> may I propose the following

that sounds like a good plan to me...how about for this PR I *only* add the gallery, and add one example along what I described above (unless there is another obvious example to add). That work?

jorisvandenbossche commented:
> that sounds like a good plan to me...how about for this PR I only add the gallery infrastructure, and add one example along what I described above (unless there is another obvious example to add). That work?

yes!

choldgraf commented:
ok @jorisvandenbossche check out the latest push for a skinnier Sphinx-Gallery deployment + one extra example that shows off ways to convert between cartopy + geopandas. Let me know what you think!

Here's what it looks like: http://predictablynoisy.com/geopandas/gallery/index.html

jorisvandenbossche commented on gallery/cartopy.py:
```diff
@@ -0,0 +1,87 @@
+# Now that we've created new Shapely objects with the CartoPy CRS,
+# we can use this to create a GeoDataFrame.
+
+df_new = gpd.GeoDataFrame(crs=crs_new.proj4_init, geometry=new_geometries)
```
And for some reason, the 'GeoDataFrame' in the rendered version links to the 'GeoDataFrame.from_file' page. Also on that page, the link to the example does not work: http://predictablynoisy.com/geopandas/api/_as_gen/geopandas.GeoDataFrame.from_file.html#geopandas.GeoDataFrame

jorisvandenbossche commented on doc/source/conf.py:
```diff
@@ -26,12 +26,33 @@
+                      'numpy': 'http://docs.scipy.org/doc/numpy/reference',
+                      'scipy': 'http://docs.scipy.org/doc/scipy/reference',
+                      'geopandas': None},
+    'backreferences_dir': 'api/_as_gen',  # Not currently used but kept for future
```
this is used now? (the comment)

choldgraf replied:
I'd say it's technically used but not super utilized. It's meant for API pages (e.g., on the API page for `from_file` it'll list all the sphinx-gallery examples that use that method/function/etc)

choldgraf replied:
I'll remove the comment since it's not technically true anymore :-)

choldgraf commented on doc/source/_static/custom.css:
```diff
@@ -0,0 +1,3 @@
+div.section a span {
```
For some reason (and I think specifically with the sphinx RTD theme) the clickable links within examples don't show up as the "link" color, but just as regular text, which makes it hard to spot as a user. This forces them to do so.

jorisvandenbossche replied:
OK, no problem, can you just add that explanation as a comment?

choldgraf commented on doc/environment.yml:
```diff
@@ -16,3 +16,4 @@ dependencies:
 - sphinx
 - sphinx_rtd_theme
 - ipython=4.0.1
+- cartopy
```
good point! though I think it'll need to be in a `pip` section because sphinx_gallery isn't available on conda

jorisvandenbossche replied:
it is on conda-forge, so it should work fine I think

choldgraf replied:
sounds good - though FYI this one won't be updated as quickly as the `pip` version (e.g., conda-forge is out of date right now)

choldgraf commented on doc/source/conf.py:
```diff
@@ -26,12 +26,33 @@
+
+# Sphinx gallery configuration
+sphinx_gallery_conf = {
+    'examples_dirs': ['../../gallery'],
```
yep for sure - just put this in another directory because examples already existed and I didn't know your intention for it. I agree examples is a better name!

choldgraf replied:
I just noticed that the `nyc_boros` file is plotting/saving files that are used later in the documentation. There's a note in there about auto-building as a part of the documentation. I can sphinx-gallerify this example and then you wouldn't need the `savefigs` in there anymore...you could just use the images that sphinx-gallery generates. LMK.

jorisvandenbossche replied:
I would leave that for later (the images are actually included in the repo, as they are also used for the README.md)

choldgraf replied:
sounds good - if you want to make this switch for future reference, all the (full-res) images will be saved to `doc/source/gallery/images`

choldgraf replied:
either way, I'll add a bit of sphinx comments to the nyc_boros.py file so that it gets built with sphinx-gallery, but the images will be saved just the same way they were before

choldgraf commented on gallery/cartopy.py:
```diff
@@ -0,0 +1,87 @@
+
+First we'll load in the data using GeoPandas.
+"""
+# sphinx_gallery_thumbnail_number = 7
```
yep, that's meant to be a visible comment

choldgraf commented on gallery/cartopy.py:
```diff
@@ -0,0 +1,87 @@
+###############################################################################
+# Note that we could have easily done this with an EPSG code like so:
+crs_epsg = cp.crs.epsg('2039')
+df_epsg = df.to_crs(epsg='2039')
```
either one should work - I'll show them side-by-side in the next commit

jorisvandenbossche replied:
I don't think either one will work, no? 
I mean, they will work for doing a coordinate transformation, but `df.to_crs(epsg='2039')` will not work to be able tot plot it on a cartopy ax ? (which is the point of this example)

choldgraf commented:
ok check out the latest push here:

http://predictablynoisy.com/geopandas/gallery/index.html

jorisvandenbossche commented on examples/nyc_boros.py:
```diff
@@ -1,36 +1,57 @@
-Time-stamp: <Tue May  6 12:17:29 EDT 2014>
+This example generates many images that are used in the documentation. See
+the `Geometric Manipulations <geometric_manipulations>` example for more
+details.
```
I think you can remove this, not of any interest for users that look at the example on the web

jorisvandenbossche replied:
Maybe, can you move the original one inside the `docs/` directory? 
The we can make this a cleaner one focused as user example

jorisvandenbossche commented on examples/nyc_boros.py:
```diff
@@ -1,36 +1,57 @@
-boros = GeoDataFrame.from_file('nybb.shp')
+path_nybb = gpd.datasets.get_path('nybb')
+boros = GeoDataFrame.from_file(path_nybb)
 boros.set_index('BoroCode', inplace=True)
```
I know it is not your code, but if we are adding it to the website: recommended is `boros = boros.set_index('BoroCode')`

jorisvandenbossche commented on examples/nyc_boros.py:
```diff
@@ -1,36 +1,57 @@
 boros.set_index('BoroCode', inplace=True)
-boros.sort()
-boros.plot()
+boros.sort_index()
```
this line is not doing anything, as it is not inplace, so `boros = boros.sort_index()` (but not sure this sorting is actually needed)

jorisvandenbossche commented on examples/nyc_boros.py:
```diff
@@ -1,36 +1,57 @@
+# Next, we'll plot the raw data
+ax = boros.plot()
 plt.xticks(rotation=90)
 plt.savefig('nyc.png', dpi=DPI, bbox_inches='tight')
```
I would remove the two lines above? (and the same for the other ones)

jorisvandenbossche commented on examples/nyc_boros.py:
```diff
@@ -1,36 +1,57 @@
+
+# Grab the limits which we'll use later
+xmin, xmax = plt.gca().get_xlim()
+ymin, ymax = plt.gca().get_ylim()
```
Do you put it here because in the next 'cell' the figure will be closed?

Would it work if you do here:

ax = boros.geometry.convex_hull.plot()

and then use `ax` in the next cell? (that would keep it cleaner, as here you do not yet need it)

jorisvandenbossche commented on examples/nyc_boros.py:
```diff
@@ -1,36 +1,57 @@
+
+##############################################################################
+# We'll generate some random dots scattered throughout our data, and will
+# use them to perform some set geometry with our boroughs. We can use
```
set -> set operation ?

jorisvandenbossche commented on examples/cartopy_convert.py:
```diff
@@ -0,0 +1,105 @@
+ax.add_geometries(new_geometries[:len(new_geometries) // 2], crs=crs_new)
+# Make the GeoPandas plot
+df_aea.iloc[len(new_geometries) // 2:].plot(column='gdp_pp', ax=ax,
+                                            vmin=0, vmax=.02)
```
I personally find this plotting of half/half the geometries with cartopy and geopandas makes the example more difficult to understand (the `[:len(new_geometries) // 2]` is additional and not necessary boilerplate to understand). 
Maybe we can plot two different things? 
Eg plot all of the countries with geopandas `.plot`, and then add eg the capitals with cartopy, or the oceans.

choldgraf commented:
@jorisvandenbossche these are all great comments - I'll give 'em a shot today on my flight to NYC and will push changes soon

choldgraf commented:
ok see latest push - I think those comments are now addressed. I went ahead and moved boros to `/doc`...I agree it'll be better to just create a more polished version. I kept all of the rST in there since it's not going to affect the code at all, could be used in a future PR, and at the least gives some comment context for it. LMK if you'd rather that be removed...

jorisvandenbossche commented on examples/cartopy_convert.py:
```diff
@@ -55,21 +55,21 @@
+# axs[0].add_geometries(df_epsg['geometry'], crs=crs_epsg,
+#                       facecolor='white', edgecolor='black')
+# # Make the GeoPandas plot
+# df_epsg.plot(ax=axs[1], color='white')
```
Is there a reason you commented this out?

choldgraf replied:
oops - no that was for debugging since I was on a plane and cartopy requires internet to do epsg lookups. just pushed it w/o these commented out

jorisvandenbossche commented:
>  I kept all of the rST in there since it's not going to affect the code at all, could be used in a future PR, and at the least gives some comment context for it. LMK if you'd rather that be removed...

Yes that is perfect!

choldgraf commented:
btw...I'm not sure why the tests aren't passing...I don't *think* I've changed anything that should have broken the codebase itself

jorisvandenbossche commented:
Don't mind the tests, that's a shapely compat issue

choldgraf commented:
sounds good! LMK if you'd like me to do anything else!

choldgraf commented:
merge conflict resolved....anything else to do here?

jorisvandenbossche commented:
I tried to test this locally, but I get "jinja2.exceptions.TemplateAssertionError: no filter named 'underline'". Any idea what it could be? (already tried to update jinja2 version, but no help)

The underline seems to occur in the autosummary template

choldgraf commented:
huh...that does sound like a jinja version error...weird. What version does it say you're using?

jorisvandenbossche commented:
I updated to 2.9.6. But does it output it somewhere when building to be sure it is using the correct one?

choldgraf commented:
hmm - not sure. Did you use pypi to install jinja? And if so, are you working with environments (e.g. with conda)? In that case, make sure that sphinx is installed in the environment you're using. Sometimes I'll update a package in one environment, but it'll actually be using the sphinx build from another environment because I haven't done `pip install sphinx` again

jorisvandenbossche commented:
Hmm, removing the pip one and force reinstalling from conda seems to solve that issue. 
But there is something wrong with the title of the generated api pages (the "<no title>" and plain text "geopandas.overlay"). Do you see that as well?

![screenshot from 2017-08-29 01-03-38](https://user-images.githubusercontent.com/1020496/29797188-10075b30-8c56-11e7-94dc-165e6465bcda.png)

choldgraf commented:
huh, it seems to look normal for me:

http://predictablynoisy.com/geopandas/api/_as_gen/geopandas.overlay.html#geopandas.overlay

this is bizarre :-/

jorisvandenbossche commented:
what version of sphinx are you using? 
I also see a lot of errors in the output like this "None:None: WARNING: toctree contains reference to document 'api/_as_gen/geopandas.GeoDataFrame' that doesn't have a title: no link will be generated" (probably related to the problem above)

choldgraf commented:
OK - I'm away from the computer now, but lemme double check all of this when I get back to it and I'll post an update tonight... That work?

jorisvandenbossche commented:
Yes, I am going to release now, so can't include it. But on readthedocs, the default landing page is 'latest' anyhow, so if I merge it later this week, it will be on there anyhow.

choldgraf commented:
just checked, I'm running:

Jinja2 (2.9.6)
Sphinx (1.5.6)
Sphinx-Gallery (0.1.12)

It feels like the jinja that your sphinx is using may not be the one that you've recently updated...not sure tho :-/

choldgraf commented:
And I think this is related to the "missing titles" problem too. Sphinx looks for titles as a line of text with `------` underneath it. The `| underline` code basically just tells jinja to add the proper number of `-----` to make it a sphinx title. So if that's broken, sphinx will also not detect the title...

jorisvandenbossche commented:
Trying this again, and updating sphinx from 1.5.2 to 1.5.6 apparently fixed the underline issue. 
But the GeoDataFrame and GeoSeries references are not yet working. 
What version of numpydoc/napoleon are you using? (not sure if that could make a difference)

choldgraf commented:
huh - that's strange...maybe the sphinx upgrade caused it to link again a different version of jinja or something?

re: the references, could you clarify what you mean by not working? Are you talking about links in examples?

jorisvandenbossche commented:
No, the example, and the links in the example are fine. The problem I see
is with the generated docstring pages. They work fine for the functions,
but not for the classes.
On the autosummary table in reference.rst, the links to the pages for
GeoDataFrame and GeoSeries also don't work, for the functions they work
nicely.

I will be offline for a week now, but after that will try further.

@jdmcbr Could you try this locally, to see whether it works for you?

Op 1 sep. 2017 6:05 p.m. schreef "Chris Holdgraf" <notifications@github.com
>:

huh - that's strange...maybe the sphinx upgrade caused it to link again a
different version of jinja or something?

re: the references, could you clarify what you mean by not working? Are you
talking about links in examples?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/geopandas/geopandas/pull/463#issuecomment-326619769>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AA-SUAwoYD5nitlsHablw_tmKL50pNisks5seCs6gaJpZM4OPGBh>
.

jdmcbr commented:
@jorisvandenbossche @choldgraf I don't see the same issue with the empty header in the API reference. I do get a bunch of warnings during build:

```
WARNING: /home/james/python/geopandas/geopandas/geodataframe.py:docstring of geopandas.GeoDataFrame:6: (SEVERE/4) Unexpected section title.

Keyword Arguments
-----------------
WARNING: /home/james/python/geopandas/doc/source/api/_as_gen/geopandas.GeoSeries.rst:11: (SEVERE/4) Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: 'source/api/_as_gen/geopandas.GeoSeries.examples'.
WARNING: /home/james/python/geopandas/geopandas/datasets/__init__.py:docstring of geopandas.datasets.get_path:4: (SEVERE/4) Unexpected section title.

Parameters
----------
WARNING: /home/james/python/geopandas/geopandas/tools/overlay.py:docstring of geopandas.overlay:8: (SEVERE/4) Unexpected section title.

Parameters
----------
/home/james/python/geopandas/geopandas/tools/overlay.py:docstring of geopandas.overlay:12: WARNING: Unexpected indentation.
/home/james/python/geopandas/geopandas/tools/overlay.py:docstring of geopandas.overlay:14: WARNING: Block quote ends without a blank line; unexpected unindent.
WARNING: /home/james/python/geopandas/geopandas/tools/overlay.py:docstring of geopandas.overlay:18: (SEVERE/4) Unexpected section title.

Returns
-------
WARNING: /home/james/python/geopandas/doc/source/api/_as_gen/geopandas.overlay.rst:11: (SEVERE/4) Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: 'source/api/_as_gen/geopandas.overlay.examples'.
WARNING: /home/james/python/geopandas/geopandas/tools/sjoin.py:docstring of geopandas.sjoin:4: (SEVERE/4) Unexpected section title.

Parameters
----------
/home/james/python/geopandas/geopandas/tools/sjoin.py:docstring of geopandas.sjoin:7: WARNING: Unexpected indentation.
/home/james/python/geopandas/geopandas/tools/sjoin.py:docstring of geopandas.sjoin:13: WARNING: Block quote ends without a blank line; unexpected unindent.
WARNING: /home/james/python/geopandas/doc/source/api/_as_gen/geopandas.sjoin.rst:11: (SEVERE/4) Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: 'source/api/_as_gen/geopandas.sjoin.examples'.
WARNING: /home/james/python/geopandas/geopandas/tools/geocoding.py:docstring of geopandas.tools.geocode:4: (SEVERE/4) Unexpected section title.

Parameters
----------
/home/james/python/geopandas/geopandas/tools/geocoding.py:docstring of geopandas.tools.geocode:7: WARNING: Unexpected indentation.
WARNING: /home/james/python/geopandas/geopandas/tools/geocoding.py:docstring of geopandas.tools.geocode:23: (SEVERE/4) Unexpected section title.

Example
-------
/home/james/python/geopandas/geopandas/tools/geocoding.py:docstring of geopandas.tools.geocode:27: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/james/python/geopandas/geopandas/tools/geocoding.py:docstring of geopandas.tools.geocode:30: WARNING: Block quote ends without a blank line; unexpected unindent.
WARNING: /home/james/python/geopandas/doc/source/api/_as_gen/geopandas.tools.geocode.rst:11: (SEVERE/4) Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: 'source/api/_as_gen/geopandas.tools.geocode.examples'.
/home/james/python/geopandas/doc/source/geocoding.rst:16: WARNING: Unknown target name: "geopy".
/home/james/python/geopandas/doc/source/reference.rst:149: WARNING: duplicate object description of GeoSeries.buffer, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:154: WARNING: duplicate object description of GeoSeries.boundary, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:159: WARNING: duplicate object description of GeoSeries.centroid, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:163: WARNING: duplicate object description of GeoSeries.convex_hull, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:170: WARNING: duplicate object description of GeoSeries.envelope, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:176: WARNING: duplicate object description of GeoSeries.simplify, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:183: WARNING: duplicate object description of GeoSeries.rotate, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:187: WARNING: duplicate object description of GeoSeries.scale, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:191: WARNING: duplicate object description of GeoSeries.skew, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:195: WARNING: duplicate object description of GeoSeries.translate, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:201: WARNING: duplicate object description of GeoSeries.unary_union, other instance in /home/james/python/geopandas/doc/source/geometric_manipulations.rst, use :noindex: for one of them
/home/james/python/geopandas/doc/source/reference.rst:209: WARNING: Unknown target name: "fiona".
/home/james/python/geopandas/doc/source/reference.rst:242: WARNING: Unknown target name: "geo_interface".
/home/james/python/geopandas/doc/source/reference.rst:264: WARNING: Unknown target name: "fiona".
/home/james/python/geopandas/doc/source/reference.rst:306: WARNING: Unknown target name: "geo_interface".
```

The inputs/outputs sections of the docs are not being properly recognized, which makes these pages makes it a little harder to read (though some of that is the formatting of the docs themselves). It looks like some linkages to examples are not working, as for the overlay example.

![overlay](https://user-images.githubusercontent.com/666108/29999937-f2c74a5a-900e-11e7-9541-ec8e9559403a.png)

choldgraf commented:
hmm, this is confusing as it's working on my build:

http://predictablynoisy.com/geopandas/api/_as_gen/geopandas.overlay.html

I think the warnings are related to the examples not showing up on the API pages. The warnings like

`InputError: [Errno 2] No such file or directory: 'source/api/_as_gen/geopandas.GeoSeries.examples'.`

Are for when sphinx looks for the rST file that generates those mini galleries (the file ending in `.examples`). So it sounds like those examples aren't being generated in the first place by sphinx-gallery.

Make sure that you've got all the necessary dependencies installed on the same environment, I've run into issues with that before.

jorisvandenbossche commented:
OK, tried this again and for some reason now it works .. (but the better!) 
The only thing I am aware of that changed is that I merged the api docs reorg PR, but I don't understand why that influence gallery.

One thing to do is to update the `make clean` command to remove all autogenerated content by sphinx-gallery. And to fix the conflict with master (since I merged the other PR, but it are no nasty conflicts).

jorisvandenbossche commented:
I only still get those warnings:

```
/home/joris/scipy/geopandas/doc/source/api/_as_gen/geopandas.GeoSeries.rst:11: SEVERE: Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: 'source/api/_as_gen/geopandas.GeoSeries.examples'.
/home/joris/scipy/geopandas/doc/source/api/_as_gen/geopandas.overlay.rst:11: SEVERE: Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: 'source/api/_as_gen/geopandas.overlay.examples'.
/home/joris/scipy/geopandas/doc/source/api/_as_gen/geopandas.sjoin.rst:11: SEVERE: Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: 'source/api/_as_gen/geopandas.sjoin.examples'.
/home/joris/scipy/geopandas/doc/source/api/_as_gen/geopandas.tools.geocode.rst:11: SEVERE: Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: 'source/api/_as_gen/geopandas.tools.geocode.examples'.
```

It seems that those are the docstrings for which there is no example yet (for the other docstring pages, I get a working link to the examples that use it). 
Is there a way to not get such warnings? (apart from writing more examples so every method is being used :-))

choldgraf commented:
ok cool - just rebased and made the changes you asked for in terms of the Makefile

I think that the errors we're seeing is because of here:

https://github.com/geopandas/geopandas/pull/463/files#diff-73befc35b23d379d3ffee5f9aaba44aaR140

the call to autosummary is what generates the `.examples` files...it'll generate one for each item listed in autosummary. The problem is that this *doesn't* happen for the methods within each class. So it then looks for an `examples` folder that doesn't exist. The errors shouldn't be a problem in terms of the build, but we should keep an eye on them in case something changes.

jdmcbr commented:
@choldgraf Is there any chance that some of the files that back your build aren't committed to this PR? I just realized I didn't see the source anywhere for the examples on your build that don't show up in my local build, but I do see the example produced from cartopy_convert.py. Which, in looking back, was maybe your intent? But in comparing your build versus mine, this was still a source of confusion for me.

choldgraf commented:
Hey @jdmcbr, which files are you talking about? The way Sphinx Gallery works is that it generates the rst files for examples when you build the documentation. But maybe you're talking about a different file?

jdmcbr commented:
@choldgraf Yes, so my source/gallery folder contains the source for the cartopy_convert example, but I don't see, for example, the source for the set_operations example. Which was maybe your intent, but was my source of confusion in comparing the results of your build with mine.

choldgraf commented:
ah that was from a previous version of this PR...that shouldn't actually be in there, and is probably just still there because the source folder wasn't cleaned out...shouldn't be a problem w/ the PR though

jorisvandenbossche commented:
Looking at this again, will try locally and if builds ok, ready to merge!

jorisvandenbossche commented:
Some small things on the actual example:

- the 4th plot does not seem to be working (the one with two subplots, the right one is empty for me)
- the last example doesn't look good anymore due to the changes in the default style (the centroids are also blue, so you can't see them)

jorisvandenbossche commented on doc/Makefile:
```diff
@@ -40,6 +40,7 @@ help:
 
 clean:
 	-rm -rf $(BUILDDIR)/*
+	-rm -rf source/api/_as_gen/*
```
Should the 'gallery' and 'api' dirs also be removed? (what do they do in other projects?)

choldgraf replied:
I usually don't remove the gallery dir since it can take a long time to generate (if the examples take a while to build). If I want that gone, then I just manually delete the folder. But LMK what you'd prefer!

choldgraf commented:
hey - just got back from out of town so picking this back up. I'll try to see what's up with the style changes!

choldgraf commented:
ok comments addressed in latest push. The non-plotting subplot was actually also a style thing. I guess the behavior of `color` changed, it was plotting *everything* as white, which made it seem like nothing was being plotted :-)

should be good to go from my end

choldgraf commented:
ping on this...I'd like to get this dangling PR closed sooner than later if possible!

jdmcbr commented:
@choldgraf I looked at this a few weeks ago, and was still getting a list of warnings, albeit a shorter list than before. I tried again this afternoon, and am still seeing the same (and yes, all dependencies installed, though I'm using pip rather than conda). Perhaps @jorisvandenbossche can check if all is building cleanly for him.

jorisvandenbossche commented:
Was just looking at it, and for me it is building fine. (there a few warnings that I mentioned above https://github.com/geopandas/geopandas/pull/463#issuecomment-329965226, but everything works fine so let's not hold that up to merge). 

So let's merge and see how it goes on readthedocs and fix in follow-up if needed!

jorisvandenbossche commented:
@choldgraf Thanks a lot for this!

choldgraf commented:
@jorisvandenbossche thanks! ping me if we need to spot-check something
